src/shutdown/umount.c

```diff
@@ -716,10 +733,44 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
                         continue;
 
                 /* Trying to umount */
-                if (umount_with_timeout(m, last_try) < 0)
+                r = umount_with_timeout(m, last_try);
+                if (r < 0)
                         n_failed++;
                 else
                         *changed = true;
+
+                /* If a mount is busy, we move it to not keep parent mount points busy.
+                 * If a mount point is not a leaf, moving it would invalidate our mount table.
+                 * More moving will occur in next iteration with a fresh mount table.
+                 *
+                 * umount_with_timeout will return EPROTO even on EBUSY due to forking.
+                 */
+                if (r != -EPROTO || !m->leaf)
+                        continue;
+
+                _cleanup_free_ char *dirname = NULL;
+
+                r = path_extract_directory(m->path, &dirname);
+                if (r < 0) {
+                        n_failed++;
+                        log_full_errno(last_try ? LOG_ERR : LOG_INFO, r, "Cannot find directory for %s: %m", m->path);
+                        continue;
+                }
+
+                if (!path_equal_or_files_same(dirname, "/run/shutdown/mounts", 0)) {
+                        char newpath[STRLEN("/run/shutdown/mounts/") + DECIMAL_STR_MAX(uint64_t) + 1];
+
+                        xsprintf(newpath, "/run/shutdown/mounts/%" PRIu64, random_u64());
+
+                        (void) mkdir_p(newpath, 0700);
```

Comment:
hmm, this assumes the mounts we move away here are dirs. they could be non-dir inodes too. Those you cannot mount on top of a dir inode. (you can mount dir inodes on dir inodes only, and non-dir on any non-dir)

hence, stat() the source first, and then either do am mkdir or a touch() to create either a dir or a regular file.

Reply:
(and yeah, i think it is worth getting this right, thinkg a root fs mounted off a loopback file whose backing fs is mounted somewhere at the leaf of the tree)

Reply:
Done. I also added error handling to calls to mkdir_p and touch_file in order to make the logs more obvious.